File: Sources/structly/05_binary_tree/HowHigh.swift
//
//  HowHigh.swift
//  MyCLI
//
//  Created by new on 3/29/26.
//

struct Pair<T: Hashable & Sendable> {
    let node: TreeNode<T>
    let height: Int
}

func howHigh(_ root: TreeNode<String>?) -> Int {
    guard let root else { return -1 }
    var stack: [Pair] = [Pair<String>(node: root, height: 0)]
    var maxHeight = 0
    while !stack.isEmpty {
        let p = stack.removeLast()
        maxHeight = max(maxHeight, p.height)
        if let right = p.node.right {
            stack.append(Pair<String>(node: right, height: p.height+1))
        }
        if let left = p.node.left {
            stack.append(Pair<String>(node: left, height: p.height+1))
        }
    }
    return maxHeight
}

File: Sources/structlyC/05_binary_tree/how_high.cpp
//
//  how_high.cpp
//  MyCLI
//
//  Created by new on 3/27/26.
//

#include <string>

class Node {
public:
    std::string val;
    Node* left;
    Node* right;

    Node(std::string initialVal) {
        val = initialVal;
        left = nullptr;
        right = nullptr;
    }
};

// number of edges
int howHigh(Node* root) {
    if (root == nullptr) {
        return -1;
    }

    if (root->left == nullptr && root->right == nullptr) {
        return 0;
    }

    return 1 + std::max(howHigh(root->left), howHigh(root->right));
}

